jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/BufferUtil.java

```diff
@@ -1099,19 +1099,24 @@ public static ByteBuffer toBuffer(Resource resource, boolean direct) throws IOEx
         if (len < 0)
             throw new IllegalArgumentException("invalid resource: " + resource + " len=" + len);
 
-        ByteBuffer buffer = direct ? BufferUtil.allocateDirect(len) : BufferUtil.allocate(len);
-
-        int pos = BufferUtil.flipToFill(buffer);
-        try (ReadableByteChannel channel = resource.newReadableByteChannel())
+        byte[] b;
+        try (InputStream is = resource.newInputStream())
         {
-            long needed = len;
-            while (needed > 0 && buffer.hasRemaining())
-            {
-                needed = needed - channel.read(buffer);
-            }
+            b = is.readAllBytes();
         }
-        BufferUtil.flipToFlush(buffer, pos);
 
+        ByteBuffer buffer;
+        if (direct)
+        {
+            buffer = BufferUtil.allocateDirect(len);
+            int pos = BufferUtil.flipToFill(buffer);
+            buffer.put(b);
+            BufferUtil.flipToFlush(buffer, pos);
+        }
```

Comment:
It feels like this should be a copy operation

Reply:
Only tests use this method. Let me see if I can move it elsewhere or improve it.